File: backend/go-gateway/internal/contracts/bootstrap.go
package contracts

// BootstrapData is the full system state returned by GET /api/bootstrap.
// The frontend MUST call this endpoint first before using any other feature.
type BootstrapData struct {
	System   SystemInfo      `json:"system"`
	Providers ProvidersSummary `json:"providers"`
	Models   ModelsSummary   `json:"models"`
	Settings SettingsSummary `json:"settings"`
	Voice    VoiceSummary    `json:"voice"`
	Features FeatureFlags    `json:"features"`
	GPU      GPUStatus       `json:"gpu"`
}

type SystemInfo struct {
	AppName    string `json:"appName"`
	AppEnv     string `json:"appEnv"`
	Version    string `json:"version"`
	APIVersion string `json:"apiVersion"`
	Uptime     string `json:"uptime"`
	Status     string `json:"status"` // "healthy" | "degraded" | "initializing"
}

type ProvidersSummary struct {
	Primary   ProviderStatus   `json:"primary"`
	Fallback  ProviderStatus   `json:"fallback"`
	Available []ProviderStatus `json:"available"`
}

type ProviderStatus struct {
	ID          string `json:"id"`
	Name        string `json:"name"`
	Status      string `json:"status"` // "available" | "provider_unavailable" | "error"
	Reason      string `json:"reason,omitempty"`
	ModelCount  int    `json:"modelCount"`
	DeviceMode  string `json:"deviceMode"` // "cloud" | "gpu" | "cpu" | "disabled"
}

type ModelsSummary struct {
	DefaultChatModel   string `json:"defaultChatModel"`
	DefaultCoderModel  string `json:"defaultCoderModel"`
	DefaultFastModel   string `json:"defaultFastModel"`
	TotalAvailable     int    `json:"totalAvailable"`
	DiscoveryEnabled   bool   `json:"discoveryEnabled"`
	LastDiscoveryTime  string `json:"lastDiscoveryTime,omitempty"`
}

type SettingsSummary struct {
	Theme           string `json:"theme"`
	Language        string `json:"language"`
	StreamingEnabled bool  `json:"streamingEnabled"`
}

type VoiceSummary struct {
	STTEnabled    bool   `json:"sttEnabled"`
	TTSEnabled    bool   `json:"ttsEnabled"`
	STTEngine     string `json:"sttEngine"`
	TTSEngine     string `json:"ttsEngine"`
	STTDeviceMode string `json:"sttDeviceMode"`
	TTSDeviceMode string `json:"ttsDeviceMode"`
}

type FeatureFlags struct {
	Chat           bool `json:"chat"`
	Voice          bool `json:"voice"`
	Vision         bool `json:"vision"`
	Terminal       bool `json:"terminal"`
	Memory         bool `json:"memory"`
	Tools          bool `json:"tools"`
	Execution      bool `json:"execution"`
	Search         bool `json:"search"`
	LocalControl   bool `json:"localControl"`
	SelfImprovement bool `json:"selfImprovement"`
	LocalLLM       bool `json:"localLlm"`
	GPUMonitor     bool `json:"gpuMonitor"`
}

// DefaultBootstrapData returns a safe default when the Python AI service is
// unavailable. The frontend shows this as a degraded-mode bootstrap.
func DefaultBootstrapData(appName, appEnv, version string, gpuRequired bool) BootstrapData {
	unavailableReason := "provider_unavailable: API key not configured"
	return BootstrapData{
		System: SystemInfo{
			AppName:    appName,
			AppEnv:     appEnv,
			Version:    version,
			APIVersion: "v1",
			Status:     "degraded",
		},
		Providers: ProvidersSummary{
			Primary: ProviderStatus{
				ID:     "nvidia_nim",
				Name:   "NVIDIA NIM",
				Status: "provider_unavailable",
				Reason: unavailableReason,
			},
			Fallback: ProviderStatus{
				ID:     "openrouter",
				Name:   "OpenRouter",
				Status: "provider_unavailable",
				Reason: unavailableReason,
			},
			Available: []ProviderStatus{},
		},
		Models: ModelsSummary{
			TotalAvailable:   0,
			DiscoveryEnabled: true,
		},
		Settings: SettingsSummary{
			Theme:            "dark",
			Language:         "en",
			StreamingEnabled: true,
		},
		Voice: VoiceSummary{
			STTEnabled: false,
			TTSEnabled: false,
		},
		Features: FeatureFlags{
			Chat:      true,
			Voice:     false,
			Vision:    false,
			Terminal:  true,
			Memory:    false,
			Tools:     true,
			Execution: false,
			Search:    false,
			LocalControl: true,
			SelfImprovement: false,
			LocalLLM:   false,
			GPUMonitor: true,
		},
		GPU: CPUFallbackGPUStatus(gpuRequired),
	}
}

File: backend/python-ai-service/app/routers/bootstrap.py
from __future__ import annotations

import time

from fastapi import APIRouter, Request

from app.core.config import get_settings
from app.core.envelopes import success
from app.core.logging import get_logger
from app.gpu.detector import GPUDetector
from app.gpu.workload_router import WorkloadRouter
from app.providers.router import ProviderRouter

logger = get_logger(__name__)
router = APIRouter()

_start_time = time.time()


@router.get("/bootstrap")
async def bootstrap(request: Request) -> dict:
    correlation_id = request.headers.get("x-correlation-id")
    settings = get_settings()
    gpu_info = GPUDetector.get_info()

    # Provider statuses
    try:
        provider_router = ProviderRouter.get()
        provider_statuses = await provider_router.get_all_statuses()
    except Exception:
        provider_statuses = []

    available_providers = [p for p in provider_statuses if p["status"] == "available"]
    primary_status = next((p for p in provider_statuses if p["id"] == settings.ai_provider_primary), {
        "id": settings.ai_provider_primary,
        "name": "NVIDIA NIM",
        "status": "provider_unavailable",
        "reason": "not initialised",
        "deviceMode": "disabled",
        "modelCount": 0,
    })
    fallback_status = next((p for p in provider_statuses if p["id"] == settings.ai_provider_fallback), {
        "id": settings.ai_provider_fallback,
        "name": "OpenRouter",
        "status": "provider_unavailable",
        "reason": "not initialised",
        "deviceMode": "disabled",
        "modelCount": 0,
    })

    # GPU status (inline — mirrors /gpu/status)
    from app.routers.gpu import _build_gpu_status
    gpu_status = _build_gpu_status()

    uptime_s = time.time() - _start_time
    system_status = "healthy" if available_providers else "degraded"

    data = {
        "system": {
            "appName": settings.app_name,
            "appEnv": settings.app_env,
            "version": "0.1.0",
            "apiVersion": "v1",
            "uptime": f"{round(uptime_s)}s",
            "status": system_status,
        },
        "providers": {
            "primary": primary_status,
            "fallback": fallback_status,
            "available": available_providers,
        },
        "models": {
            "defaultChatModel": "",
            "defaultCoderModel": "",
            "defaultFastModel": "",
            "totalAvailable": 0,
            "discoveryEnabled": settings.model_discovery_enabled,
        },
        "settings": {
            "theme": "dark",
            "language": "en",
            "streamingEnabled": True,
        },
        "voice": {
            "sttEnabled": settings.stt_enabled,
            "ttsEnabled": settings.tts_enabled,
            "sttEngine": settings.stt_engine,
            "ttsEngine": settings.tts_engine,
            "sttDeviceMode": "gpu" if (settings.stt_gpu_enabled and gpu_info.cuda_available) else "cpu",
            "ttsDeviceMode": "cpu",
        },
        "features": {
            "chat": True,
            "voice": settings.stt_enabled or settings.tts_enabled,
            "vision": settings.vision_enabled,
            "terminal": True,
            "memory": settings.faiss_enabled,
            "tools": True,
            "execution": settings.sandbox_enabled,
            "search": settings.web_search_enabled,
            "localControl": settings.local_pc_control_enabled,
            "selfImprovement": settings.self_improvement_enabled,
            "localLlm": settings.local_llm_enabled,
            "gpuMonitor": settings.gpu_enabled,
        },
        "gpu": gpu_status,
    }

    return success(data, correlation_id)
